### ISSUE TITLE
linalg.trace_dot - comparison of GPys implementation with an implementation based on Magnus & Neuendecker

### ISSUE THREAD

**Issue by ingmarschuster** [SheffieldML/GPy#161]
I've been implementing my own trace_dot function (based on the formula in Magnus & Neuendeckers "Matrix differential calculus") and later found GPys implementation.
I did a little comparision, not sure what to make of it, so I'm posting it for discussion. GPys trace_dot is faster, but also has larger error when assuming that actually doing np.trace(np.dot(a,b)) is ground truth (I'm not sure this assumption is correct). Also, the Magnus-Neuendecker implementation should actually be equivalent to the implementation in GPy as far as I can see - but isn't numerically.

The %timeit magic in the code below requires IPython.

```
def trace_dot_magnus(a, b):
    return np.dot(a.T.flat, b.flat)

def trace_dot_gpy(a, b):
    """
    Efficiently compute the trace of the matrix product of a and b
    """
    return np.sum(a * b)

def abs_error_trace_dot(a, b, func):
    return func(a,b) - np.trace(np.dot(a,b))

a = [np.random.rand(3,3) for _ in range(100)]

gpy    = [abs_error_trace_dot(a[i], a[i+1], trace_dot_gpy   ) for i in range(len(a) - 1)]
magnus = [abs_error_trace_dot(a[i], a[i+1], trace_dot_magnus) for i in range(len(a) - 1)]
print("GPy abs error", np.sum(gpy), "average", np.average(gpy))
print("Magnus abs error", np.sum(magnus), "average", np.average(magnus))

%timeit [trace_dot_gpy(a[i], a[i+1])    for i in range(len(a) - 1)]
%timeit [trace_dot_magnus(a[i], a[i+1]) for i in range(len(a) - 1)]
```

**Comment by mikecroucher** [comment]
The GPy implementation was wrong! This might not have been picked up because it doesn't appear to be used anywhere in the project right now. We can see that it's wrong using 2x2 matrices in Sympy:

```python
from sympy import *
var('a:h')

#Compute the trace the slow way
A = Matrix([[a,b],[c,d]])
B = Matrix([[e,f],[g,h]])
C = A*B
trace = C.trace()

#Compute the trace the original GPy way
D = A.multiply_elementwise(B)
GPy_trace = sum(D)

#Corrected implementation
E = A.multiply_elementwise(B.T)
fixed_trace = sum(E)
```

If you run this, you'll see that the correct trace is `a*e + b*g + c*f + d*h` but the GPy implementation had it as `a*e + b*f + c*g + d*h`. Taking the transpose of B in the GPy version fixes this.

We can, however, do better using einsum. Observe the following timings

```python
import numpy as np
from numpy.core.umath_tests import inner1d

def trace_dot_magnus(a, b):
    return np.dot(a.T.flat, b.flat)

def trace_dot_gpy(a, b):
    return np.sum(a * b)

def trace_dot_fast(a,b):
    return np.einsum('ij,ji->', a, b)

def abs_error_trace_dot(a, b, func):
    return func(a,b) - np.trace(np.dot(a,b))

a = [np.random.rand(3,3) for _ in range(100)]

gpy    = [abs_error_trace_dot(a[i], a[i+1], trace_dot_gpy   ) for i in range(len(a) - 1)]
magnus = [abs_error_trace_dot(a[i], a[i+1], trace_dot_magnus) for i in range(len(a) - 1)]
fast = [abs_error_trace_dot(a[i], a[i+1], trace_dot_fast) for i in range(len(a) - 1)]

print("GPy abs error", np.sum(gpy), "average", np.average(gpy))
print("Magnus abs error", np.sum(magnus), "average", np.average(magnus))
print("Fast abs error", np.sum(fast), "average", np.average(fast))

print('GPy')
%timeit [trace_dot_gpy(a[i], a[i+1])    for i in range(len(a) - 1)]
print('magnus')
%timeit [trace_dot_magnus(a[i], a[i+1]) for i in range(len(a) - 1)]
print('fast')
%timeit [trace_dot_fast(a[i], a[i+1]) for i in range(len(a) - 1)]
```

On my machine, I get
```
GPy abs error -4.06994775209 average -0.0411105833544
Magnus abs error -2.22044604925e-16 average -2.24287479722e-18
Fast abs error 0.0 average 0.0
GPy
1000 loops, best of 3: 265 µs per loop
magnus
1000 loops, best of 3: 949 µs per loop
fast
10000 loops, best of 3: 129 µs per loop
```

This idea comes from http://stackoverflow.com/questions/18854425/what-is-the-best-way-to-compute-the-trace-of-a-matrix-product-in-numpy and I will submit a PR with it in as a fix.

**Comment by ingmarschuster** [comment]
Ha, great! I didn't think of einsum, this is the best solution of course.

**Comment by mikecroucher** [comment]
Fixed in commit c307b98
